src/Symfony/Component/Messenger/Worker.php

```diff
@@ -191,6 +191,20 @@ private function dispatchEvent($event)
 
     private function shouldRetry(\Throwable $e, Envelope $envelope, RetryStrategyInterface $retryStrategy): bool
     {
+        if ($e instanceof HandlerFailedException) {
+            //checks if ALL Exceptions are an instance of UnrecoverableExceptionInterface
+            $shouldNotRetry = array_reduce(
```

Comment:
a foreach would be more common + would allow breaking the loop earlier

Reply:
i adjusted the commit and expanded the test